service/history/shard/context_impl.go

```diff
@@ -367,7 +367,20 @@ func (s *ContextImpl) UpdateQueueAckLevel(
 			ClusterAckLevel: make(map[string]int64),
 		}
 	}
-	s.shardInfo.QueueAckLevels[category.ID()].AckLevel = convertTaskKeyToAckLevel(category.Type(), ackLevel)
+	persistenceAckLevel := convertTaskKeyToPersistenceAckLevel(category.Type(), ackLevel)
+	s.shardInfo.QueueAckLevels[category.ID()].AckLevel = persistenceAckLevel
+
+	// if cluster ack level is less than the overall ack level, update cluster ack level
+	// as well to prevent loading too many tombstones if the cluster ack level is used later
+	// this may happen when adding back a removed cluster or rolling back the change for using
+	// single queue in timer/transfer queue processor
+	if queueAckLevel, ok := s.shardInfo.QueueAckLevels[category.ID()]; ok {
+		for clusterName, persistenceClusterAckLevel := range queueAckLevel.ClusterAckLevel {
+			if persistenceClusterAckLevel < persistenceAckLevel {
+				queueAckLevel.ClusterAckLevel[clusterName] = persistenceAckLevel
+			}
+		}
+	}
```

Comment:
should we do this only when singleCursor is enabled?

Reply:
i guess if singleCursor is not enabled, then this is just no-op?

Reply:
In most cases yes, it's no-op when single cursor is disabled. 
But, like explained in the comment, when removing clusters, the ack level associated with that cluster is still stored in shard info. If later that cluster is added back, we will start from a very old ack level for that cluster and encountering tombstone issue when loading tasks.

Anyway, those logic will be gone once multi-cursor is ready as we won't have cluster ack level any more.